crates/resolver-tests/src/lib.rs

```diff
@@ -274,22 +417,47 @@ pub struct SatResolver(Rc<RefCell<SatResolverInner>>);
 
 struct SatResolverInner {
     solver: varisat::Solver<'static>,
+    old_root_vars: Vec<varisat::Var>,
     var_for_is_packages_used: HashMap<PackageId, varisat::Var>,
-    by_name: HashMap<&'static str, Vec<PackageId>>,
+    var_for_is_packages_features_used: HashMap<PackageId, HashMap<InternedString, varisat::Var>>,
+    by_name: HashMap<InternedString, Vec<Summary>>,
 }
 
 impl SatResolver {
     pub fn new(registry: &[Summary]) -> Self {
-        let mut cnf = varisat::CnfFormula::new();
+        let mut solver = varisat::Solver::new();
```

Comment:
The `Solver` is used directly without creating a `CnfFormula`, since otherwise the variable counter is reset when creating the `Solver` and the root package variable conflicts with the first registry package.

Reply:
As I recall the CNF was used because of a performance problem when adding clauses directly to `Solver`. Although that may have been fixed in the intervening years. It seems like there's a fair bit of complexity caused by adding clauses in side `sat_resolve`, on the other hand there's some good code sharing thanks to your new `add_pkg_clauses`.

I wonder if we could avoid the complexity in `sat_resolve` by deciding that the root package cannot have features. Its dependencies can activate transitive features, which may activate transitive optional dependencies. But, the exact process of going from user requested features to a list of dependencies for the root crate has to be done before calling `sat_resolve`. That's obviously less flexible. But are there real bugs that this simplification could not test? What are your thoughts?

I will continue to review when I have time tomorrow. Possibly during my office hours. Depending on who attends. If you attend we could discuss this work face to face!

Reply:
> As I recall the CNF was used because of a performance problem when adding clauses directly to `Solver`. Although that may have been fixed in the intervening years. It seems like there's a fair bit of complexity caused by adding clauses in side `sat_resolve`, on the other hand there's some good code sharing thanks to your new `add_pkg_clauses`.

I didn't see any performance regression in the existing prop tests, since `SatResolver::new()` is  only called one time and then reused.

> I wonder if we could avoid the complexity in `sat_resolve` by deciding that the root package cannot have features. Its dependencies can activate transitive features, which may activate transitive optional dependencies. But, the exact process of going from user requested features to a list of dependencies for the root crate has to be done before calling `sat_resolve`. That's obviously less flexible. But are there real bugs that this simplification could not test? What are your thoughts?

We cannot replace clauses in `sat_resolve` by simple assumptions, because a required dependency for the root package is represented by a `OR` clause with all compatible packages.

What about the following example:
```toml
# Cargo.toml
[package]
name = "root"

[dependencies]
indexmap = { version = "2.5.0", optional = true }
# ...
# other deps that may activate `indexmap`
# ...
```
```sh
# command-line
cargo c -F 'indexmap?/serde'
```
Assuming `--all-features` from the cli is not enough, since the argument introduces a new feature not specified anywhere. Keeping the feature clauses for the root package can help reusing code in this case.

Additionally this example doesn't work with vendoring, since the generated `Cargo.lock` doesn't contains `serde` (the version resolution for `serde` is done on the fly when executing the command, but the result is not saved anywhere).

I don't think there are any test for this, since all existing tests assume `--all-features` for the root package.

Reply:
> I didn't see any performance regression in the existing prop tests, since SatResolver::new() is only called one time and then reused.

Each prop test is called hundreds of times each with a distinct registry, `SatResolver::new()` is called once per generated registry. I was also not able to reproduce the performance problem, so I guess it's been solved at some point in the past few years. So all good.

> We cannot replace clauses in `sat_resolve` by simple assumptions, because a required dependency for the root package is represented by a `OR` clause with all compatible packages.

Right, but that's not new.  The old code also had to deal with  a dependency from the root being an `OR`  clause. That is what the old code was using `this_call`/`matches` to do. It seems that the old code successfully added a clause to the solver without clobbering the internal numbering system  or hitting the conflict you described.

> I don't think there are any test for this, since all existing tests assume `--all-features` for the root package.

Yes, I don't think there any tests for this behavior. If I remember correctly, the ability for the command line to  turn on any arbitrary dependencies arbitrary feature was accidentally introduced.  Whether we are locked in by backwards compatibility, or whether we can consider it a bug, is being discussed in other places.

The question at hand is... Imagine we  had a prop test that completely covered all possible `root` crates that only have required features, so it will test the registry with `indexmap = { version = "2.5.0" }` and `indexmap = { version = "2.5.0", features =["serde"]}`. (Prop tests never give  guarantees, but if we've been running them long enough we have a probabilistic statement of coverage.) Do we think it's possible that a user  will find a bug that uses command line `-F`  that cannot be reproduced by the prop test?

Reply:
> The question at hand is... Imagine we had a prop test that completely covered all possible root crates that only have required features, so it will test the registry with indexmap = { version = "2.5.0" } and indexmap = { version = "2.5.0", features = ["serde"] }.

Yes this should work: testing my above example is equivalent to testing `root1`, `root2`, `root3`, `root4` like this:
```toml
[package]
name = "dep"
version = "1.0.0"

[dependencies]
indexmap = { version = "2.5.0", optional = true }
# ...

[features]
cli1 = []
cli2 = ["dep:indexmap"]
cli3 = ["indexmap/serde"]
cli4 = ["indexmap?/serde"]
```
```toml
[package]
name = "root1"

[dependencies]
dep = { version = "1.0.0", features = ["cli1"]
```
```toml
[package]
name = "root2"

[dependencies]
dep = { version = "1.0.0", features = ["cli2"]
```
```toml
[package]
name = "root3"

[dependencies]
dep = { version = "1.0.0", features = ["cli3"]
```
```toml
[package]
name = "root4"

[dependencies]
dep = { version = "1.0.0", features = ["cli4"]
```

So we can limit the root package to have only non-optional dependencies and no features, so that `sat_resolve` don't need to handle cli features.

> It seems that the old code successfully added a clause to the solver without clobbering the internal numbering system or hitting the conflict you described.

The solver allows reusing a variable not bound by any clause after adding a cnf formula. With the following registry and dependencies, the old code gives the same variable value `3` for the dep `b` and the `root` package:
```
let reg = registry(vec![
    pkg!(("a", "1.5.0")),
    pkg!(("a", "1.6.0")),
    pkg!(("b", "1.5.0")),
    pkg!(("c", "1.4.0")),
]);
let deps = vec![dep_req("a", "1.5.0"), dep("b")];
```
This is not a problem for the old code since it doesn't keep a reference of the root variable inside the `SatResolve`.